opentracing/span.py

```diff
@@ -228,3 +223,58 @@ def log(self, **kwargs):
         if 'timestamp' in kwargs:
             timestamp = kwargs['timestamp']
         return self.log_kv(key_values, timestamp)
+
+
+class ContinuationMixin(object):
+    """
+    Continuation mixin that provides additional primitives to a `Span` class,
+    so that it can be deactivated and resumed later during in-process
+    propagation.
+
+    If the application needs to defer work that should be part of the same
+    `Span`, the `ContinuationMixin` provides a `capture` method that increments
+    a counter to the span so that it can be deactivated without finishing. In
+    another asynchronous executor and/or thread, it can be resumed.
+    """
+    def capture(self):
+        """
+        Capture the `Span` tracing session as well as any 3rd-party execution
+        context of interest. After calling this method, the `Span` may be used
+        later in a closure or callback function where it may be resumed and
+        reactivated using the tracer `active_span_source.make_active()`.
+
+        IMPORTANT: the caller MUST activate and then deactivate the `Span`
+        otherwise it will never automatically finish. That is, calling
+        `capture()` increments a refcount that must be decremented somewhere
+        else.
+        """
+        pass
+
+    def deactivate(self):
+        """
+        Mark the end of the active period for the current `Span` in this
+        asynchronous executor and/or thread. When the last refcount has been
+        decremented through `deactivate()`, it is automatically finished.
+
+        NOTE: Calling `deactivate()` more than once on a single `Span` instance
+        leads to undefined behavior.
+        """
+        self.finish()
+
+
+class Span(ContinuationMixin, BaseSpan):
```

Comment:
@bhs @palazzem Upon further research while preparing test cases and test scenarios, I came upon a few things to add here:

When it comes to the `to_restore` part, the only place where I can actually see that happening is when using a thread pool (through `concurrent.futures`), as at least both Gevent/asyncio keep their context-locals new for each new task, no no need to restore anything. 

And for the mentioned thread pool, observe threads *do not* get suspended/reactivated (as I suspect happens for Java+Futures+ExecutorService or C#+async). So in theory you would always get a thread that *doesn't need* to keep its previous state (`Continuation` objects would still be needed, though, of course).

That being said:

1) Most of the time the user wouldn't simply need to do a capture/activate cycle, for both the async frameworks and the multi threaded approaches. And when the user needs it, `Continuation` can be available, and any slightly usability issue around it is probably fine at that point. 
2) If we were to provide an `ActiveSpan` just as Java does (say, for sanity purposes on restoring *any* previous data), how much would it actually impact the most common scenarios? Since `ActiveSpan` is basically a proxy, and the main difference is the `finish` vs `deactivate` methods (which are meant to be used through a context manager) is not such a big burden on the API IHMO.